File: nomenclator/src/main/java/flowmanager/nomenclator/model/WorkItem.java
package flowmanager.nomenclator.model;

import jakarta.persistence.*;
import lombok.*;

import java.time.LocalDate;
import java.util.ArrayList;
import java.util.List;

@Data
@AllArgsConstructor
@NoArgsConstructor
@EqualsAndHashCode(exclude = {
        "project",
        "comments",
        "reporter",
        "assignees",
        "parent",
        "children"
})
@Builder
@Entity
@Table(name = "work_item")
public class WorkItem {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Integer id;

    @Column(nullable = false)
    private String title;

    @Column(nullable = false)
    private String description;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private ItemType itemType;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private Status status;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private Severity severity;

    @Column(nullable = false)
    private LocalDate createdAt;

    @Column
    private LocalDate dueDate;

    @ManyToOne
    @JoinColumn(name = "project_id", nullable = false)
    private Project project;

    @Builder.Default
    @OneToMany(mappedBy = "workItem")
    private List<Comment> comments = new ArrayList<>();

    @ManyToOne
    @JoinColumn(name = "reporter_id", nullable = false)
    private User reporter;

    @Builder.Default
    @ManyToMany
    @JoinTable(
            name = "work_item_assignment",
            joinColumns = @JoinColumn(name = "work_item_id"),
            inverseJoinColumns = @JoinColumn(name = "user_id")
    )
    private List<User> assignees = new ArrayList<>();

    @ManyToOne
    @JoinColumn(name = "parent_id")
    private WorkItem parent;

    @Builder.Default
    @OneToMany(mappedBy = "parent")
    private List<WorkItem> children = new ArrayList<>();
}

File: frontend/app/types/workItem.ts
import type { ItemType, Severity, Status } from "./enums"
import type { UserSummaryDto } from "./user"
import type { ProjectSummaryDto } from "./project"
import type { CommentResponseWorkItemDto } from "./comment"

export interface WorkItemCreateDto {
  title: string
  description: string
  itemType: ItemType
  severity: Severity
  projectId: number
  parentId?: number
  dueDate?: string
  assigneesIds?: number[]
}

export interface WorkItemUpdateDto {
  title?: string
  description?: string
  status?: Status
  severity?: Severity
  dueDate?: string
  assigneesIds?: number[]
}

export interface WorkItemSummaryDto {
  id: number
  title: string
  description: string
  itemType: ItemType
  status: Status
  severity: Severity
  createdAt: string
  dueDate: string | null
  projectId: number
  reporter: UserSummaryDto
  assignees: UserSummaryDto[]
}

export interface WorkItemResponseDto {
  id: number
  title: string
  description: string
  itemType: ItemType
  status: Status
  severity: Severity
  createdAt: string
  dueDate: string | null
  project: ProjectSummaryDto
  comments: CommentResponseWorkItemDto[]
  reporter: UserSummaryDto
  assignees: UserSummaryDto[]
  parent: WorkItemSummaryDto | null
  children: WorkItemSummaryDto[]
}
